nixos/modules/services/web-apps/invoiceplane.nix

```diff
@@ -0,0 +1,206 @@
+{ config, pkgs, lib, ... }:
+
+let
+  inherit (lib) mkDefault mkEnableOption mkForce mkIf mkMerge mkOption types;
+  inherit (lib) any attrValues concatMapStringsSep flatten literalExpression;
+  inherit (lib) filterAttrs mapAttrs mapAttrs' mapAttrsToList nameValuePair optional optionalAttrs optionalString;
+
+  cfg = migrateOldAttrs config.services.invoiceplane;
+  eachSite = cfg.sites;
+  user = "invoiceplane";
+  webserver = config.services.${cfg.webserver};
+  stateDir = hostName: "/var/lib/invoiceplane/${hostName}";
+
+  # Migrate config.services.invoiceplane.<hostName> to config.services.invoiceplane.sites.<hostName>
+  oldSites = filterAttrs (o: _: o != "sites" && o != "webserver");
+  migrateOldAttrs = cfg: cfg // { sites = cfg.sites // oldSites cfg; };
+
+  pkg = hostName: cfg: pkgs.stdenv.mkDerivation rec {
+    pname = "invoiceplane-${hostName}";
+    version = src.version;
+    src = cfg.package;
+
+    installPhase = ''
+      mkdir -p $out
+      cp -r * $out/
+    '';
+  };
+
+  siteOpts = { lib, name, ... }:
+    {
+      options = {
+
+        enable = mkEnableOption "InvoicePlane web application.";
+
+        package = mkOption {
```

Comment:
I'm not sure this option as any value here... does it?

Reply:
Do you mean the enable or the package option? I use enable so I can start the service like this:
```
services.invoiceplane.sites."localhost".enable = true;
```
The package option is used in other web apps as well.

Reply:
I mean the `package` option.

> The package option is used in other web apps as well.

This really isn't a justification... and you didn't answer my question :confused:

Reply:
Hm isn't this `package` option useful if I want to use a custom overlay package instead of the unmodified `invoiceplane` package?

Reply:
If you want to use a custom overlay package instead of the unmodified `invoiceplane` package you use an overlay. Overlays overwrite packages in your nixpkgs.

```nix
overlays = [
  (self: super: {
    invoiceplane = <something that will override invoiceplane everywhere in your OS, including the invoiceplane module>
  })
];
```

I think most people don't understand what overlays are really for, and this accounts for a large portion of the `package` options in NixOS :man_shrugging:

Reply:
So you would recommend to remove this option? I'm okay with that ;)

Reply:
It isn't that I'm against a `package` option... but I do notice that having a `package` option everywhere seems to keep more and more of our users in the dark about what overlays actually are and when they should be used.

In this particular case I don't think it is overly useful. If a security fix ever comes out we need to update immediately and push that out, there are no variants of the package, etc...

Ultimately, up to you.